### ISSUE TITLE
Issues in Aggregate public key set by federation block 

### ISSUE THREAD

**Issue by Naviabheeman** [chaintope/tapyrus-core#237]
The following problems are found while setting  aggregate pub key in federation params using xfield:
1.  When a node restarts with "-reindex" after the chain's aggregate pub key is updated by a federation block, it makes the ferderation params inconsistent. 
before restart
`[{'025700236c2890233592fcef262f4520d22af9160e3d9705855140eb2aa06c35d3': 0}, {'03831a69b8009833ab5b0326012eaf489bfea35a7321b1ca15b11d88131423fafc': 12}, {'02bf2027c8455800c7626542219e6208b5fe787483689f1391d6d443ec85673ecf': 25}`
and after restart it recovers to:
`[{'025700236c2890233592fcef262f4520d22af9160e3d9705855140eb2aa06c35d3': 0}, {'03831a69b8009833ab5b0326012eaf489bfea35a7321b1ca15b11d88131423fafc': 12}, {'02bf2027c8455800c7626542219e6208b5fe787483689f1391d6d443ec85673ecf': 25},
{'03831a69b8009833ab5b0326012eaf489bfea35a7321b1ca15b11d88131423fafc': 12}, {'02bf2027c8455800c7626542219e6208b5fe787483689f1391d6d443ec85673ecf': 25}])`
Genesis block aggpubkey is loaded at startup. After this during the "-reindex" phase the next two aggpubkey changes are added. Then when the tip is  updated until the latest block is reached using ConnectBlock() two more aggpubkey changes are added to the list. 

- To fix this we can make ReadAggregatePubkey() check the list before adding a new entry. 

2. It's not possible to add a new node to a network after the aggpubkey change. I.e. The p2p protocol used HEADERS message to send headers first and then the complete blocks. When a block header with aggpubkey change is present in the list, all headers following this should be checked using the new aggpubkey. But this is not happening now. So any new node would assume that the headers are invalid and disconnect.

- To fix this we need to make the p2p protocol aware of xfield in block headers.  The receiving node should process the xfield in block header during the headers message processing and not wait for the entire block. But federation params should be updated only when the block is accepted to the tip.

3. Restarting a node after an xfield change causes VerifyDB to fail. So that node doesn't start. it should be started using "-reindex"
`2022-11-28T05:47:44.662717Z [0%]...ERROR: ReadBlockFromDisk: ReadBlockFromDisk: bad-proof, Proof verification failed (code 16) nHeight = 23
2022-11-28T05:47:44.663413Z ERROR: VerifyDB(): *** ReadBlockFromDisk failed at 23, hash=4fc79ddc6897f07a741605b9d9d8b00bbbc5d1502578083161a811c763ae73c2`

**Comment by Naviabheeman** [comment]
To fix 2 and 3 we need to make all the aggregate pubkeys persistent. We can use the block index in level DB. It stores the block location along with the block header as a key value pair. This bitcoin page states the exact specification of data stored in the index DB: https://en.bitcoin.it/wiki/Bitcoin_Core_0.11_(ch_2):_Data_Storage

We can add a new key value pair for aggpubkey :
```
Key : X1 (for xfield type 1)
Value: <ser_aggpubkey><height>
```

**Comment by azuchi** [comment]
> Key : X(for xfield)<xfield_type>

In this case, if there are multiple aggregated pubkey updates, won't the key be duplicated and only the latest aggregated pubkey will be retrieved?

**Comment by azuchi** [comment]
```
Key : X(for xfield)<xfield_type><height as 4 bytes>
Value:<xfield_value><block hash if need>
```

In this case, a range search from`X<type_value>00000000` to `X<type_value>ffffffff` can list registered values. 

I'm not particularly picky, so something like SQLite is acceptable if it is easy to implement, just like Bitcoin Core's wallet has changed from Berkeley DB to SQLite.

**Comment by Naviabheeman** [comment]
> In this case, if there are multiple aggregated pubkey updates, won't the key be duplicated and only the latest aggregated pubkey will be retrieved?
yes!

I checked levels documentation and it does not limit the size of value. But it states that the performance is not good at 1000,000 bytes. Since our data in value is fixed size and can we just keep one db entry per xfield, one key and the serialised list of xfield changes as the value?

```
Key : <xfield_type>
Value: <size> <xfield_value><height><block hash> <xfield_value><height><block hash> etc
```
level db is storing data in sorted order of keys. So for aggpubkey, "1" as the key so that the xfield data will be at the beginning of the sorted keys. Even if we add more xfields, all xfield data will be together at the beginning of the db.

**Comment by Naviabheeman** [comment]
v0.5 fix does not let the blockchain reorg to remove a federation block i.e block with xfield changing aggpubkey. In the master branch, we need to decide if this should be allowed for 1. _aggpubkey_ and 2. _maxblocksize_

**Comment by Naviabheeman** [comment]
Reorg behaviour:

consider this scenario:

B0 -> B1 ....-> B30-> .............-> B40 (Federation block) -> B41 ....
  ........................   | -> B31 ............ -> B42 -> ....

If the longer chain is B42 then there is a reorg and B40 is lost. But the aggpubey is not removed from federation params. This may result in blocks being rejected from the new chain after B40 due to proof verification failure.  When the node is restarted with "-reindex" the global aggregatepubkey list would be corrected. The chain with B42 is the longer chain and the node would continue to use genesis aggpubkey to verify block proof. 

`Invalidateblock` and `submitblock` RPCs do not allow changing a block before the federation block. But with the P2P protocol, reorg is not prevented. This is the observed  behaviour on the master branch. This should be the same in v0.5.2 too.

**Comment by Naviabheeman** [comment]
issues fixed in 0.5.2 and master branch.
